### PR TITLE
Patch primary key

### PR DESCRIPTION
**Pull request by KhristenkoYura**
is not added primary key in scheme for distributed index

### CODE REVIEW

**Comment by samdark** [comment]
Can you elaborate a bit about distributed index and why it's not there now...

**Comment by KhristenkoYura** [comment]
http://sphinxsearch.com/docs/current/distributed.html

Run this test: https://github.com/KhristenkoYura/yii2-sphinx/commit/e3448baf1ae1e859c64aa257d24146a566907952 ( https://travis-ci.org/KhristenkoYura/yii2-sphinx/jobs/75919186#L248 )
IndexSchema::$primaryKey is null only for distributed index

File: Schema.php [hunk 1]
```diff
@@ -449,7 +449,35 @@ protected function getColumnPhpType($column)
      */
     protected function findColumns($index)
     {
-        $sql = 'DESCRIBE ' . $this->quoteSimpleIndexName($index->name);
+        $columns = $this->getIndexColumns($index->name);
+        if (!empty($columns[0]['Agent'])) {
```
Comment (njspok) [inline Schema.php]: Please, add comment about distributed index special case

**Comment by pastuhov** [comment]
+1

**Comment by njspok** [comment]
+1

**Comment by PaVeL-Ekt** [comment]
+1

File: Schema.php (hunk 1 above)
Comment (KhristenkoYura) [inline Schema.php]: Напишите здесь комментарий, и я его добавлю

**Comment by samdark** [comment]
@pastuhov, @njspok, @PaVeL-Ekt have you verified that it works?

**Comment by sa-kirich** [comment]
@samdark yes, it works!

**Comment by samdark** [comment]
@klimov-paul, @cebe  any comments before merge?

File: Schema.php [hunk 2]
```diff
@@ -449,7 +449,35 @@ protected function getColumnPhpType($column)
...
+    }
+
+
+    /**
+     * Return columns for index $name
+     * @param $name string index name
+     * @return array|bool
+     * @throws \Exception
+     */
+    protected function getIndexColumns($name) {
```
Comment (klimov-paul) [inline Schema.php]: `getIndexColumn` is not a good name for this method. Especially comparing with `findColumns`.
Also code style is broken.

File: tests/SchemaTest.php [hunk 3]
```diff
@@ -30,6 +30,18 @@ public function testGetIndexSchemas()
         }
     }
 
+    public function testGetPrimaryKeySchema() {
```
Comment (klimov-paul) [inline tests/SchemaTest.php]: Code style is broken.

File: tests/SchemaTest.php [hunk 4]
```diff
@@ -30,6 +30,18 @@ public function testGetIndexSchemas()
...
+        $indexes = $this->getConnection()->schema->getIndexSchemas();
+        foreach($indexes as $index) {
+            foreach($index->columns as $column) {
+                if ($column->isPrimaryKey) {
+                    $this->assertEquals($index->primaryKey, $column->name);
```
Comment (klimov-paul) [inline tests/SchemaTest.php]: Test seems to be useless.
 Does it fails without thi patch?

File: Schema.php [hunk 5]
```diff
@@ -449,7 +449,35 @@ protected function getColumnPhpType($column)
      */
     protected function findColumns($index)
     {
-        $sql = 'DESCRIBE ' . $this->quoteSimpleIndexName($index->name);
+        $columns = $this->getIndexColumns($index->name);
+        if (!empty($columns[0]['Agent'])) {
+            $columns = $this->getIndexColumns($columns[0]['Agent']);
```
Comment (klimov-paul) [inline Schema.php]: THe original code was recursive; agent of the destributed index can contain its own 'Agent', which should be traversed. This patch breaks the recursion.

**Comment by njspok** [comment]
@samdark Yes it works

File: tests/SchemaTest.php (hunk 3 above)
Comment (KhristenkoYura) [inline tests/SchemaTest.php]: Why?

File: tests/SchemaTest.php (hunk 4 above)
Comment (KhristenkoYura) [inline tests/SchemaTest.php]: So you can talk about all the tests. If you want, I can remove the method

File: tests/SchemaTest.php (hunk 3 above)
Comment (samdark) [inline tests/SchemaTest.php]: `{` should be at the next line as far as I can see...

File: tests/SchemaTest.php (hunk 4 above)
Comment (samdark) [inline tests/SchemaTest.php]: Usually a test is added for the thing that is fixed. @klimov-paul is trying to tell that in this case it's not for the code that was changed. Is it true?

**Comment by samdark** [comment]
@klimov-paul does it look OK now?

File: tests/SchemaTest.php (hunk 4 above)
Comment (klimov-paul) [inline tests/SchemaTest.php]: Extra test never will be superfluous.
Howevere, does this test expose the reported issue? It seems it does not.

File: Schema.php [hunk 6]
```diff
@@ -449,31 +449,34 @@ protected function getColumnPhpType($column)
...
-            foreach ($columns as $info) {
-                $column = $this->loadColumnSchema($info);
-                $index->columns[$column->name] = $column;
-                if ($column->isPrimaryKey) {
-                    $index->primaryKey = $column->name;
-                }
+            // for distributed index
+            list($next, $indexName) = !empty($columns[0]['Agent'])
+                ? array(true, $columns[0]['Agent'])
+                : array(false, $indexName);
```
Comment (klimov-paul) [inline Schema.php]: An odd construction: why compact an array, which should be split right away?

Comment (KhristenkoYura) [inline Schema.php]: It can be done through "if else", but a "list" looks laconically

File: tests/SchemaTest.php (hunk 3 above)
Comment (KhristenkoYura) [inline tests/SchemaTest.php]: Correct https://github.com/KhristenkoYura/yii2-sphinx/commit/8f7783807887fc9107603017eee4635e98876c63

File: Schema.php (hunk 6 above)
Comment (samdark) [inline Schema.php]: `if else` is better in this case, I think.

Comment (KhristenkoYura) [inline Schema.php]: Which is best?
1.
 ```php
            list($next, $indexName) = !empty($columns[0]['Agent'])
                ? array(true, $columns[0]['Agent'])
                : array(false, $indexName);
```
2.
 ```php
            if (!empty($columns[0]['Agent'])) {
                $next = true;
                $indexName = $columns[0]['Agent'];
            } else {
                $next = false;
            }
```

I like the first option.

Comment (samdark) [inline Schema.php]:
```php
$next = false;
if (!empty($columns[0]['Agent'])) {
    $next = true;
    $indexName = $columns[0]['Agent'];
}
```

Comment (KhristenkoYura) [inline Schema.php]: https://github.com/KhristenkoYura/yii2-sphinx/commit/c94dcbae550aa23cc9ebe0da5f52a8d1b651b7db

**Comment by samdark** [comment]
Good. So the only question is about test. Is it testing the issue fixed or something else?

**Comment by KhristenkoYura** [comment]
In the test I just corrected the code style.
If there are more wishes I can be corrected.

**Comment by samdark** [comment]
Nope. You've added `SchemaTest::testGetPrimaryKeySchema()` method and we're not sure it fails w/o the fix provided in this pull request.

**Comment by KhristenkoYura** [comment]
Corrected test.
You can specify, what do you doubt?

**Comment by klimov-paul** [comment]
Related to #45

**Comment by samdark** [comment]
@klimov-paul any issues? I think it's ready to be merged...

**Comment by klimov-paul** [comment]
Th is not correct fix - too complex.
Stand by.

**Comment by klimov-paul** [comment]
Resolved by commit 65cc2eb11ea54720225b4e7d8503e19cae61aa96

**Comment by samdark** [comment]
Nice :)

**Pull request closed without merging**